File: problems/0347/solution.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
using namespace std;

class Solution
{
public:
    vector<int> topKFrequent(vector<int> &nums, int k)
    {
        unordered_map<int, int> count = {};
        int m = 1;
        for (const auto &n : nums)
        {
            count[n] += 1;
            m = max(count[n], m);
        }
        vector<vector<int>> bucket(m);
        for (const auto &n : count)
        {
            bucket[n.second - 1].push_back(n.first);
        }
        vector<int> result;
        for (int i = m - 1; i >= 0; i--)
        {
            for (const auto &n : bucket[i])
            {
                result.push_back(n);
            }
            if (result.size() == k)
            {
                return result;
            }
        }
        return {};
    }
};

int main()
{
    vector<int> nums = {1, 1, 1, 2, 2, 3};
    int k = 2;
    Solution sol;
    vector<int> result = sol.topKFrequent(nums, k);
    for (const auto &n : result)
    {
        cout << n << ' ';
    }
}

File: problems/0347/solution.py
from typing import List
from collections import Counter
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
    #     myDict = {}
    #     for n in nums:
    #         if n not in myDict:
    #             myDict[n] = 1
    #         myDict[n] += 1
    #     ret = sorted([(n,count) for n, count in myDict.items()], key=lambda item: item[1],reverse=True)
    #     return [n for n, count in ret][0:k]
        counts = Counter(nums)
        bucket = [[] for _ in range(max(counts.values()))]
        for n, count in counts.items():
            bucket[count-1].append(n)
        result = []
        for tmp in reversed(bucket):
            result += tmp
            if len(result) == k:
                return result


if __name__ == "__main__":
    nums = [1,2,2,3,3,3]
    k = 2
    print(Solution().topKFrequent(nums=nums, k = k))